lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -74,6 +75,118 @@ static LogicalResult verifyCircuitOp(CircuitOp &circuit) {
     return failure();
   }
 
+  // Store a mapping of defname strings to the first external module
+  // that defines it.
+  llvm::StringMap<FExtModuleOp> defnameMap;
+
+  // Verify external modules.
+  for (auto &op : *circuit.getBody()) {
+    if (auto extModule = dyn_cast<FExtModuleOp>(op)) {
+      if (auto defname = extModule.defnameAttr()) {
+        // Check that this extmodule's defname does not conflict with
+        // the symbol name of any module.
+        auto collidingModule = circuit.lookupSymbol(defname.getValue());
+        if (isa_and_nonnull<FModuleOp>(collidingModule)) {
+          auto diag =
+              op.emitOpError()
+              << "attribute 'defname' with value " << defname
+              << " conflicts with the name of another module in the circuit";
+          diag.attachNote(collidingModule->getLoc())
+              << "previous module declared here";
+          return failure();
+        }
+
+        // Find an optional extmodule with a defname collision. Update
+        // the defnameMap if this is the first extmodule with that
+        // defname or if the current extmodule takes no parameters and
+        // the collision does. The latter condition improves later
+        // extmodule verification as checking against a parameterless
+        // module is stricter.
+        auto collidingExtModuleOpt =
+            Optional<FExtModuleOp>(llvm::NoneType::None);
```

Comment:
ops are nullable, so I'd just declare this as `FExtModuleOp` instead of using Optional explicitly.

Reply:
Updated to use null and not option. (This will take some getting used to...)